### DISCUSSION TITLE
Efficient inference for streaming samples in tree ensembles

### DISCUSSION THREAD

**Discussion by meghana-madhyastha** [online-ml/river#727] [Ideas]
I recently wrote a paper on this topic (published in KDD 2021):  https://dl.acm.org/doi/10.1145/3447548.3467368. Here, we assume the tree ensemble model doesn't fit in memory (it supports in-memory models as well, but benefits are greater for external memory models). The test data points streamed in. During inference, we selectively access tree nodes. The tree nodes are re-organized a-priori using heuristics to minimize the number of disk I/Os performed. This is more efficient than reading the whole model because in the streaming case, only log(n) number of nodes are read. The workflow is as follows. We read a trained tree ensemble from sklearn/xgboost/other. We then rearrange the nodes in the tree and memory map them to support selectively accessing the tree nodes for inference.  I implemented this in C++ along with a python package (with python bindings) in a separate repository, but I would like to add integrate this with river. I briefly discussed this with Max Halford over email correspondences and he expressed interest. Max suggested that I post my proposal with technical details here. 

Here I describe some of the main optimizations my paper introduced and a high level strategy for integration with river. I would love to hear your feedback.

 In our method, we make the following modifications for efficiency. (Note: In 1) and 2) below the trained tree nodes can either be reorganized or the nodes can stored in the layouts described as and when new nodes are added to the ensemble. )

 1.  Interleave root nodes and nodes of depth 1 for all trees (this is for cache efficiency purposes that I'll be happy to explain in more detail over call). This might be a little tricky to encorporate directly into river. This is because we cannot inherit from EnsembleMixin since we are storing all the nodes of the ensemble together (and not storing them explicitly as separate trees since we are interleaving some nodes in the tree). One solution is to create another ensemble class that doesn't inherit from EnsembleMixin and store all the tree nodes together. 
2.  Weighted depth first search: This is easy to add to river (and can be directly incorporated into all the existing tree types). Here, instead of storing the nodes in depth first search order, we store either left or right node first depending on cardinality. This way we keep popular paths together. 
3.   Selective Access: For single observation inference (eg if we just want to perform inference without updating the model), we don't need to read the entire model. Nodes can be read on demand. In our code, we provide this functionality with mmap (memory mapping the file). That way, we don't need to make any algorithmic changes, but just need to mmap the saved model on disk. This can be done with apache arrow. To my understanding the model in river is saved in a pandas dataframe (that can be converted to an apache arrow format and vice versa). If you're open to accepting other serialization formats, this is a pretty straightforward addition. Otherwise, I can look at python mmap to see if it can be integrated.

Overall I see two ways in which my work can be integrated and I would like to seek your opinion on the preferred way.

 1.  A separate ensemble class with all the modifications I mentioned above that can work with different trees (hoeffding/extremely fast decision tree/isop tree/etc). This way I won't have to modify much of the existing code and my optimizations can be easily added.
2.   I can start by adding optimization 2) above to existing trees. I can also try to add a serialization format that supports selective access with minimal/no changes to the existing training/inference API.
Thanks,
Meghana

**Comment by MaxHalford** [comment]
Hello!

> I recently wrote a paper on this topic (published in KDD 2021): https://dl.acm.org/doi/10.1145/3447548.3467368.

Congrats 👏

>  I implemented this in C++ along with a python package (with python bindings) in a separate repository

Can you provide a link to said repository?

> To my understanding the model in river is saved in a pandas dataframe

What led you to that understanding? River models are intended to be pickled, that's all. Also it feels like Apache Arrow as a dependency would be too burdensome.

> I can start by adding optimization 2) above to existing trees

I wouldn't rush into it adding this to River just yet. I would suggest writing some code to make your proposal with one kind of tree first. Essentially, I would write a minimal reproducible example that works. Then we can see if your proposal does indeed bring any benefit. 

As I told you by email, your proposal sounds interesting. But I have to admit that it sounds a tad complicated. To make it into River, you'll have to convince with a working prototype. I would really focus on that right now.
